### ISSUE TITLE
ElasticsearchIndicesClient#getMapping(GetMappingRequest) throws on reading response when mapping contains an object property

### ISSUE THREAD

**Issue by sothawo** [elastic/elasticsearch-java#45]
Using version 7.15.2.

When an index mapping contains a property that is an object, 
`co.elastic.clients.elasticsearch.indices.ElasticsearchIndicesClient#getMapping(co.elastic.clients.elasticsearch.indices.GetMappingRequest)` crashes on parsing the result.

This can be tested with the following code:

```java
class Name {
	String first;
	String last;

	public Name(String first, String last) {
		this.first = first;
		this.last = last;
	}
       // getter+setter
}

class Person {
	String id;
	Name name;

	public Person(String id, Name name) {
		this.id = id;
		this.name = name;
	}
       // getter+setter
}

ElasticsearchClient client = ... // setup the ElasticsearchClient
String index = "testindex";
Person person = new Person("42", new Name("Ford", "Prefect"));
client.index(b -> b.index(index).id(person.id).document(person));
GetMappingResponse getMappingResponse = client.indices().getMapping(mrb -> mrb.index(index));

```
This will produce the following error:
```
jakarta.json.stream.JsonParsingException: Property 'type' not found

	at co.elastic.clients.json.JsonpUtils.lookAheadFieldValue(JsonpUtils.java:139)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:128)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:95)
	at co.elastic.clients.json.BuildFunctionDeserializer.deserialize(BuildFunctionDeserializer.java:42)
	at co.elastic.clients.json.JsonpDeserializer$LazyDeserializer.deserialize(JsonpDeserializer.java:205)
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:102)
	at co.elastic.clients.json.JsonpDeserializer$StringMapDeserializer.deserialize(JsonpDeserializer.java:461)
	at co.elastic.clients.json.JsonpDeserializer$StringMapDeserializer.deserialize(JsonpDeserializer.java:447)
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:102)
	at co.elastic.clients.json.ObjectDeserializer$FieldObjectDeserializer.deserialize(ObjectDeserializer.java:68)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:122)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:95)
	at co.elastic.clients.json.BuildFunctionDeserializer.deserialize(BuildFunctionDeserializer.java:42)
	at co.elastic.clients.json.JsonpDeserializer$LazyDeserializer.deserialize(JsonpDeserializer.java:205)
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:102)
	at co.elastic.clients.json.ObjectDeserializer$FieldObjectDeserializer.deserialize(ObjectDeserializer.java:68)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:122)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:95)
	at co.elastic.clients.json.BuildFunctionDeserializer.deserialize(BuildFunctionDeserializer.java:42)
	at co.elastic.clients.json.JsonpDeserializer$LazyDeserializer.deserialize(JsonpDeserializer.java:205)
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:102)
	at co.elastic.clients.base.DictionaryResponse.lambda$setupDictionaryResponseDeserializer$0(DictionaryResponse.java:148)
	at co.elastic.clients.json.ObjectDeserializer.parseUnknownField(ObjectDeserializer.java:150)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:120)
	at co.elastic.clients.json.ObjectDeserializer.deserialize(ObjectDeserializer.java:95)
	at co.elastic.clients.json.BuildFunctionDeserializer.deserialize(BuildFunctionDeserializer.java:42)
	at co.elastic.clients.json.JsonpDeserializer$LazyDeserializer.deserialize(JsonpDeserializer.java:205)
	at co.elastic.clients.json.JsonpDeserializer.deserialize(JsonpDeserializer.java:102)
	at co.elastic.clients.base.rest_client.RestClientTransport.getHighLevelResponse(RestClientTransport.java:242)
	at co.elastic.clients.base.rest_client.RestClientTransport.performRequest(RestClientTransport.java:104)
	at co.elastic.clients.elasticsearch.indices.ElasticsearchIndicesClient.getMapping(ElasticsearchIndicesClient.java:1044)
	at co.elastic.clients.elasticsearch.indices.ElasticsearchIndicesClient.getMapping(ElasticsearchIndicesClient.java:1061)
```
The mapping returned from the server:

```json
{
  "testindex" : {
    "mappings" : {
      "properties" : {
        "id" : {
          "type" : "text",
          "fields" : {
            "keyword" : {
              "type" : "keyword",
              "ignore_above" : 256
            }
          }
        },
        "name" : {
          "properties" : {
            "first" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "last" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            }
          }
        }
      }
    }
  }
}
```

I debugged through the code; when reading the properties, the parser comes to the `name` entry and then on finding a START_OBJECT tries to get the `type` property of the object, basically expecting something like
```json
{
    "type": "object",   <===  Elasticsearch does not send a type entry here
    "properties": {
      "first": {
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword",
            "ignore_above": 256
          }
        }
      },
      "last": {
        "type": "text",
        "fields": {
          "keyword": {
            "type": "keyword",
            "ignore_above": 256
          }
        }
      }
    }
```

Even when the mapping is stored explicitly in Elasticsearch with the type object, it is not returned on getting the mapping.

**Comment by pullannagari** [comment]
Hi @sothawo , I am new to contributing to open source - can I pick this issue and start working?

**Comment by sothawo** [comment]
I am not maintaining this project, just reporting this issue. @swallez surely can tell you more.

**Comment by swallez** [comment]
@sothawo this was a bug in the [API spec](https://github.com/elastic/elasticsearch-specification) that is used to generate the client code. This has been fixed in the upcoming 7.16 (I have checked that it works).

The exception thrown for missing properties has also been improved in 7.16 to report not only the property name but also the class name.

Since the API spec is a huge effort to formalize the Elasticsearch API, we may keep on encountering this kind of issues.  I've added a way to have [scoped disabling of required properties checks](https://github.com/elastic/elasticsearch-java/commit/9867b65e9f3c734ac89acf732741cca1f6593258) in order to not be blocked by this kind of issue. Yes, it's kind of a hack, but a least allows to move forward without having to wait for the next release of the client just for that 😉

**Comment by swallez** [comment]
@pullannagari as you may have seen from the comment above, it has been solved already!

It's nice for you to want to contribute to open source projects. Now this project is a bit uncommon in that it has two really distinct parts:
- the Elasticsearch API implementation, in the `co.elastic.clients.elasticsearch` package which is entirely generated from the [API specification](https://github.com/elastic/elasticsearch-specification). Contribution there has to be done through that specification, and so isn't straightforward. We do however appreciate reports of discrepancies with what Elasticsearch actually expects in requests and returns in responses.
- the supporting framework in other packages, tests, and everything else, which are "normal" handwritten code and where code contributions are welcome.

**Comment by sothawo** [comment]
@swallez I am currently adding the support for this client to Spring Data Elasticsearch and am  running my integration tests against ES with both the `RestHighLevelClient` and the new client. For tests that fail I have a JUnit `@DisabledIf("newElasticsearchClient")` flag with which I exclude the test that show errors from currently not yet supported features. This way I can keep integrating and when a new version comes out I can check which of these tests then will pass.

I'm building this in a way so that in the next version of Spring Data Elasticsearch (4.4, about in a half year) the default client will still be the `RestHighLevelClient` and the new client will be optional. The idea/plan is for the next big release that is based on Java 17 with Spring 6 support (end of next year) to use the new client and drop the `RestHighLevelClient` support then.

**Comment by pullannagari** [comment]
> @pullannagari as you may have seen from the comment above, it has been solved already!
> 
> It's nice for you to want to contribute to open source projects. Now this project is a bit uncommon in that it has two really distinct parts:
> 
> * the Elasticsearch API implementation, in the `co.elastic.clients.elasticsearch` package which is entirely generated from the [API specification](https://github.com/elastic/elasticsearch-specification). Contribution there has to be done through that specification, and so isn't straightforward. We do however appreciate reports of discrepancies with what Elasticsearch actually expects in requests and returns in responses.
> * the supporting framework in other packages, tests, and everything else, which are "normal" handwritten code and where code contributions are welcome.

Thanks @swallez for taking time to look into my request, will look the advised topics further.

**Comment by sothawo** [comment]
@swallez I just tried it with the 7.16.0-SNAPSHOT version from the elasticsearch snapshot repository and still get the exception

**Comment by sothawo** [comment]
> The exception thrown for missing properties has also been improved in 7.16 to report not only the property name but also the class name.

I don't see that either in 7.16, even if I build 7.16 locally directly from the code and use that jar.

**Comment by swallez** [comment]
Finally nailed it. ES indeed sends a `type` field (the discriminant for polymorphic deserialization) for all property types except `"object"`, so we have to consider it as the default value is `type` is missing.

Fixed, verified with a [unit test](https://github.com/elastic/elasticsearch-java/blob/0af73c231572683985601175ced98567a2fe2bf4/java-client/src/test/java/co/elastic/clients/elasticsearch/model/VariantsTest.java#L113) and an [integration test](https://github.com/elastic/elasticsearch-java/blob/0af73c231572683985601175ced98567a2fe2bf4/java-client/src/test/java/co/elastic/clients/elasticsearch/end_to_end/RequestTest.java#L353). Please cross-check.

**Comment by sothawo** [comment]
I can confirm that this works with the 7.16.0-SNAPSHOT.

**Comment by swallez** [comment]
@LordOfTheHelmet please open a new issue as this is a different issue from this one. Thanks!
